host/xtest/xtest_4000.c

```diff
@@ -5027,3 +5039,112 @@ static void xtest_tee_test_4010(ADBG_Case_t *c)
 out:
 	TEEC_CloseSession(&session);
 }
+
+static void xtest_tee_test_4011(ADBG_Case_t *c)
+{
+	TEEC_Session s = { 0 };
+	size_t key_size = 512;
+	TEE_ObjectHandle key;
+	TEE_OperationHandle ops;
+	TEE_OperationHandle opv;
+	TEE_OperationHandle ope;
+	TEE_OperationHandle opd;
+	uint32_t ret_orig;
+	uint8_t in[TEE_SHA1_HASH_SIZE];
+	uint8_t out[1024];
+	uint8_t tmp[1024];
+	size_t out_size;
+	size_t tmp_size;
+	size_t n;
+	size_t m;
+	size_t i = 0;
+
+	/* Setup session, initialize message to sign, create a keypair */
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, xtest_teec_open_session(&s,
+			&crypt_user_ta_uuid, NULL, &ret_orig)))
+		return;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, ta_crypt_cmd_random_number_generate(c,
+			&s, in, sizeof(in))))
+		goto out;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, ta_crypt_cmd_allocate_transient_object(
+			c, &s, TEE_TYPE_RSA_KEYPAIR, key_size, &key)))
+		goto out;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, ta_crypt_cmd_generate_key(c, &s,
+			key, key_size, NULL, 0)))
+		goto out;
+
+	/* Allocate operations for sign, verify, encrypt and decrypt */
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, ta_crypt_cmd_allocate_operation(c, &s,
+			&ops, TEE_ALG_RSASSA_PKCS1_V1_5_SHA1, TEE_MODE_SIGN,
+			key_size)))
+		goto out;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, ta_crypt_cmd_allocate_operation(c, &s,
+			&opv, TEE_ALG_RSASSA_PKCS1_V1_5_SHA1, TEE_MODE_VERIFY,
+			key_size)))
+		goto out;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, ta_crypt_cmd_allocate_operation(c, &s,
+			&ope, TEE_ALG_RSA_NOPAD, TEE_MODE_ENCRYPT, key_size)))
+		goto out;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c, ta_crypt_cmd_allocate_operation(c, &s,
+			&opd, TEE_ALG_RSA_NOPAD, TEE_MODE_DECRYPT, key_size)))
+		goto out;
+
+	/* Assign the keypair to all operations */
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+		ta_crypt_cmd_set_operation_key(c, &s, ops, key)))
+		goto out;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+		ta_crypt_cmd_set_operation_key(c, &s, opv, key)))
+		goto out;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+		ta_crypt_cmd_set_operation_key(c, &s, ope, key)))
+		goto out;
+	if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+		ta_crypt_cmd_set_operation_key(c, &s, opd, key)))
+		goto out;
+
+	for (i = 0; i < 9; i++) {
+		Do_ADBG_Log("Iteration %zu", i);
+
+		out_size = sizeof(out);
+		if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+			ta_crypt_cmd_asymmetric_sign(c, &s, ops, NULL, 0,
+				in, sizeof(in), out, &out_size)))
+			goto out;
+
+		if (!ADBG_EXPECT_TEEC_SUCCESS(c,
+			ta_crypt_cmd_asymmetric_verify(c, &s, opv, NULL, 0,
+				in, sizeof(in), out, out_size)))
+			goto out;
+
+		tmp_size = sizeof(tmp);
```

Comment:
From here on and down to the decryption, it would be nice to have a comment saying what is really going on. It's not obvious when reading the code. Nothing big, just a sentence or two.